lib/spack/spack/spec.py

```diff
@@ -1922,15 +2096,38 @@ def parse_compiler(self, text):
         return self.compiler()
 
 
-    def spec(self):
+    def spec_by_hash(self):
+        self.expect(ID)
+
+        specs = spack.installed_db.query()
+        matches = [spec for spec in specs if
+                   spec.dag_hash()[:len(self.token.value)] == self.token.value]
+
+        if not matches:
+            tty.die("%s does not match any installed packages." %self.token.value)
+
+        if len(matches) != 1:
+            tty.error("%s matches multiple installed packages:" %self.token.value)
+            print
+            display_specs(matches, long=True)
+            print
+            print "You can either:"
+            print "  a) Use a more specific hash, or"
+            print "  b) Specify the package by name."
+            sys.exit(1)
```

Comment:
Parser should raise some type of ParseException (e.g. an `AmbiguousHashError` that extends `ParseError`) rather than printing all this and exiting.  Give the caller a chance to handle the error.  Also the parser shouldn't be responsible for any user interaction.